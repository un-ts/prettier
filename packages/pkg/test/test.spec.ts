import path from 'node:path'
import { fileURLToPath } from 'node:url'

import _ from 'lodash'
import { type Options, format } from 'prettier'

import pkg1 from './fixtures/fixture1.json'
import pkg2 from './fixtures/fixture2.json'
import pkg3 from './fixtures/fixture3.json'

import PkgPlugin from 'prettier-plugin-pkg'

const pkgs = [pkg1, pkg2, pkg3]

const createFixture = (index: 0 | 1 = 0) => {
  const pkg: Record<string, unknown> = pkgs[index]
  return _.shuffle(Object.keys(pkg)).reduce(
    (acc, key) =>
      Object.assign(acc, {
        [key]: pkg[key],
      }),
    {},
  )
}

const JSON_STRINGIFY = 'json-stringify'

const _dirname =
  typeof __dirname === 'undefined'
    ? path.dirname(fileURLToPath(import.meta.url))
    : __dirname

test('randomize', async () => {
  const input = JSON.stringify(createFixture(), null, 2)
  const output = await format(input, {
    filepath: path.join(_dirname, 'package.json'),
    parser: JSON_STRINGIFY,
    plugins: [PkgPlugin],
  })

  expect(output).toMatchSnapshot()
})

test('preprocess', async () => {
  const input = JSON.stringify(createFixture(1), null, 2)
  const output = await format(input, {
    filepath: path.join('package.json'),
    parser: JSON_STRINGIFY,
    plugins: [PkgPlugin],
    pluginSearchDirs: false,
    preprocess(content: string) {
      const { version, repository } = JSON.parse(content) as typeof pkg1
      return { repository, version }
    },
  } as Options)

  expect(output).toMatchSnapshot()
})

test('not package.json', async () => {
  const fixture = { version: 'batman', name: 'joker' }
  const input = JSON.stringify(fixture, null, 2)
  const output = await format(input, {
    filepath: 'batman.json',
    parser: JSON_STRINGIFY,
    plugins: [PkgPlugin],
  })

  expect(input.trim()).toBe(output.trim())
})

test('broken json', () => {
  const broken = `{
  "name": "prettier-plugin-pkg",
  "batman": {]
}`

  return expect(() =>
    format(broken, {
      filepath: 'broken.json',
      parser: JSON_STRINGIFY,
      plugins: [PkgPlugin],
    }),
  ).rejects.toThrow()
})
