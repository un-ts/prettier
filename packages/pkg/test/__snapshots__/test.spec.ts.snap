// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preprocess 1`] = `
"{
  "name": "prettier-plugin-pkg",
  "version": "0.0.0",
  "description": "An opinionated package.json formatter plugin for Prettier",
  "repository": "git@github.com/un-ts/prettier.git",
  "homepage": "https://github.com/un-ts/prettier/tree/master/packages/pkg",
  "author": "JounQin (https://www.1stG.me) <admin@1stg.me>",
  "license": "MPL-2.0",
  "engines": {
    "node": ">= 8.0.0"
  },
  "main": "lib/cjs",
  "module": "lib/esm",
  "es2015": "lib/es2015",
  "files": [
    "lib",
    "rules",
    "README.md",
    "LICENSE"
  ],
  "keywords": [
    "package",
    "package.json",
    "pkg",
    "plugin",
    "prettier",
    "prettier-plugin"
  ],
  "scripts": {
    "bootstrap": "yarn build && tsc --module commonjs --outDir lib/cjs",
    "build": "ts-node -O '{\\"module\\":\\"commonjs\\"}' build"
  },
  "peerDependencies": {
    "prettier": "^1.18.2"
  }
}
"
`;

exports[`randomize 1`] = `
"{
  "name": "prettier-plugin-pkg",
  "version": "0.0.0",
  "description": "An opinionated package.json formatter plugin for Prettier",
  "repository": "git@github.com/un-ts/prettier.git",
  "homepage": "https://github.com/un-ts/prettier/tree/master/packages/pkg",
  "author": "JounQin (https://www.1stG.me) <admin@1stg.me>",
  "license": "MPL-2.0",
  "engines": {
    "node": ">= 8.0.0"
  },
  "main": "lib/cjs",
  "module": "lib/esm",
  "es2015": "lib/es2015",
  "files": [
    "lib",
    "rules"
  ],
  "keywords": [
    "package",
    "package.json",
    "pkg",
    "plugin",
    "prettier",
    "prettier-plugin"
  ],
  "peerDependencies": {
    "prettier": "^1.18.2"
  },
  "resolutions": {
    "@babel/core": "^7.9.0",
    "@babel/preset-env": "^7.12.1",
    "eslint-plugin-prettier": "^3.1.2",
    "imagemin-gifsicle": "^6.0.1",
    "prettier": "^2.0.4",
    "typescript": "^3.8.3"
  },
  "commitlint": {
    "extends": [
      "@1stg"
    ]
  },
  "eslintIgnore": [
    "CHANGELOG.md",
    "lib",
    "!.*.js"
  ]
}
"
`;
